### PR TITLE
Simulate climber

### LINKED ISSUE #20: Simulate climber
A simulation of the climber may help the operator practice climbing and it is a great way to check code.

### LINKED ISSUE #38: Climber Safety features
The Climber should only work at the "endgame", and shouldn't used when we know it can harm the robot or the environment.

### PR DESCRIPTION
**Pull request by adamv17**

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -1,35 +1,34 @@
 package frc.robot;
 
+import edu.wpi.first.math.util.Units;
+
```
Comment (barel13) [inline src/main/java/frc/robot/Constants.java]: You can remove the import statement

File: src/main/java/frc/robot/Ports.java [hunk 2]
```diff
@@ -27,5 +28,10 @@
         public static final boolean RIGHT_SENSOR_PHASE = true;
         public static final boolean IS_RIGHT_INVERTED = true;
 
+        public static final int JOYSTICK_PORT = 0;
+
+        public static final int ENCODER_A_CHANNEL = 0;
+        public static final int ENCODER_B_CHANNEL = 1;
```
Comment (barel13) [inline src/main/java/frc/robot/Ports.java]: Please add comment that you only use this Constants for simulation

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 3]
```diff
@@ -3,18 +3,84 @@
...
+
     private static Climber INSTANCE = null;
+
     private final WPI_TalonFX leftMotor = new WPI_TalonFX(Ports.Climber.LEFT);
     private final WPI_TalonFX rightMotor = new WPI_TalonFX(Ports.Climber.RIGHT);
+
     private final UnitModel unitModel = new UnitModel(Constants.Climber.TICKS_PER_RAD);
 
+    private final ArmFeedforward feedforward = new ArmFeedforward(0, 0, 0, 0);
+    private final PIDController m_controller = new PIDController(1.5, 0, 0);
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: Add the coefficients to Constants

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 4]
```diff
@@ -3,18 +3,84 @@
...
+    private final DCMotor m_armGearbox = DCMotor.getFalcon500(2);
+
+    private final SingleJointedArmSim m_armSim =
+            new SingleJointedArmSim(
+                    m_armGearbox,
+                    Constants.Climber.GEAR_RATIO,
+                    SingleJointedArmSim.estimateMOI(Constants.Climber.ARM_LENGTH, Constants.Climber.ARM_MASS),
+                    Constants.Climber.ARM_LENGTH,
+                    Units.degreesToRadians(-75),
+                    Units.degreesToRadians(255),
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: Constants

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 5]
```diff
@@ -86,43 +153,75 @@ public static Climber getInstance() {
         return INSTANCE;
     }
 
+    // TODO: add units
+
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: So add units!!

File: src/main/java/frc/robot/subsystems/climber/commands/JoystickClimb.java [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 import edu.wpi.first.wpilibj2.command.CommandBase;
 import frc.robot.Constants;
+import frc.robot.Robot;
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/commands/JoystickClimb.java]: Remove unused import

**Review by barel13 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Constants.java [hunk 7]
```diff
@@ -1,35 +1,42 @@
...
+        public static final double ARM_ENCODER_DIST_PER_PULSE = 2.0 * Math.PI / 2048;
+        public static final double ARM_MASS = 5.0; // Kilograms
+        public static final double ARM_LENGTH = 0.792; // [m]
+
+        public static final int MAX_ANGLE = 255;
+        public static final int MIN_ANGLE = -75;
+
+        public static final int AUTONOMY_TIME = 15;
+
+        public static final double STOP_CLIMBER = 249.5; // [s]
```
Comment (barel13) [inline src/main/java/frc/robot/Constants.java]: I think it's too long (it's 4 minutes)

File: src/main/java/frc/robot/Constants.java [hunk 8]
```diff
@@ -1,35 +1,42 @@
...
+
+        public static final double ARM_ENCODER_DIST_PER_PULSE = 2.0 * Math.PI / 2048;
+        public static final double ARM_MASS = 5.0; // Kilograms
+        public static final double ARM_LENGTH = 0.792; // [m]
+
+        public static final int MAX_ANGLE = 255;
+        public static final int MIN_ANGLE = -75;
+
+        public static final int AUTONOMY_TIME = 15;
+
```
Comment (barel13) [inline src/main/java/frc/robot/Constants.java]: Add units

File: src/main/java/frc/robot/subsystems/climber/commands/JoystickClimb.java [hunk 9]
```diff
@@ -16,25 +17,39 @@
...
+    public void initialize() {
+        if (Robot.isReal() && !Robot.debug)
+            isReady = climber.getFPGATime() > Constants.Climber.AUTONOMY_TIME;
+    }
+
     @Override
     public void execute() {
-        double climbVelocity = Utils.deadband(joystickOutput.getAsDouble(), Constants.Climber.THRESHOLD) * Constants.Climber.MAX_VELOCITY;
-        climber.setVelocity(climbVelocity);
+        if (isReady) {
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/commands/JoystickClimb.java]: You can delete the if statement

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 10]
```diff
@@ -86,43 +139,90 @@ public static Climber getInstance() {
...
+    public double getFPGATime() {
+        return Timer.getFPGATimestamp();
+    }
+
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: You can delete this function

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 11]
```diff
@@ -3,18 +3,85 @@
...
 import frc.robot.subsystems.UnitModel;
 
+
 public class Climber extends SubsystemBase {
+
     private static Climber INSTANCE = null;
+
     private final WPI_TalonFX leftMotor = new WPI_TalonFX(Ports.Climber.LEFT);
     private final WPI_TalonFX rightMotor = new WPI_TalonFX(Ports.Climber.RIGHT);
+    private final Solenoid stopper = new Solenoid(PneumaticsModuleType.CTREPCM, Ports.Climber.STOPPER);
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: Use main/follower,aux...

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 12]
```diff
@@ -86,43 +139,90 @@ public static Climber getInstance() {
...
     public void setVelocity(double velocity) {
-        int ticks100ms = unitModel.toTicks100ms(velocity);
-        rightMotor.set(ControlMode.Velocity, ticks100ms);
-        leftMotor.set(ControlMode.Velocity, ticks100ms);
+        if (Robot.isSimulation()) {
+            double volts = m_controller.calculate(getVelocity(), velocity);
+            rightMotor.setVoltage(volts + feedforward.calculate(0, 0));
+        } else {
+            int tick100ms = unitModel.toTicks100ms(velocity);
+            rightMotor.set(ControlMode.Velocity, tick100ms);
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: You can inline it

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 13]
```diff
@@ -86,43 +139,90 @@ public static Climber getInstance() {
...
+
     /**
      * @return get motors velocity.
      */
     public double getVelocity() {
-        return unitModel.toVelocity(rightMotor.getSelectedSensorVelocity());
+        if (Robot.isSimulation()) {
+            return m_encoderSim.getRate();
+        }
+        return unitModel.toVelocity(leftMotor.getSelectedSensorVelocity());
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: right is the main motor

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 14]
```diff
@@ -86,43 +139,90 @@ public static Climber getInstance() {
...
+            rightMotor.setVoltage(volts + feedforward.calculate(0, 0));
+        } else {
+            int tick100ms = unitModel.toTicks100ms(velocity);
+            rightMotor.set(ControlMode.Velocity, tick100ms);
+        }
     }
 
     /**
-     * @return get motors position.
+     * @return get motors position. [ticks/rad}
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: [ticks/rad]

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 15]
```diff
@@ -86,43 +139,90 @@ public static Climber getInstance() {
...
+
+    public boolean isStopperEngaged() {
+        return stopper.get();
+    }
+
+    public void setStopperMode(boolean engaged) {
+        stopper.set(!engaged);
+    }
+
+    public void toggle() {
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: maybe toggleStopper?

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 16]
```diff
@@ -86,43 +139,90 @@ public static Climber getInstance() {
...
+        if (isStopperEngaged()) {
+            setStopperMode(false);
+        } else {
+            setStopperMode(true);
+        }
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: You can simplify this if statement

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 17]
```diff
@@ -64,18 +123,12 @@ public Climber() {
         /*
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java:126]: config cruise velocity and max acceleration

File: src/main/java/frc/robot/Constants.java [hunk 18]
```diff
@@ -1,35 +1,42 @@
...
-        public static final double P_RIGHT_POSITION = 0;
-        public static final double I_RIGHT_POSITION = 0;
-        public static final double D_RIGHT_POSITION = 0;
+        public static final double F_FORWARD_S = 0;
+        public static final double F_FORWARD_COS = 0;
+        public static final double F_FORWARD_V = 0;
+        public static final double F_FORWARD_A = 0;
 
+        public static final double MAX_ACCELERATION = 0;
+        public static final double MIN_ACCELERATION = 0;
```
Comment (barel13) [inline src/main/java/frc/robot/Constants.java]: should be cruise velocity

File: src/main/java/frc/robot/Leds.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
+package frc.robot;
```
Comment (barel13) [inline src/main/java/frc/robot/Leds.java]: The class should be in its own branch

File: src/main/java/frc/robot/Ports.java [hunk 20]
```diff
@@ -20,12 +21,21 @@
...
         public static final int RIGHT = 0;
+        public static final int STOPPER = 0;
 
         public static final boolean LEFT_SENSOR_PHASE = true;
         public static final boolean IS_LEFT_INVERTED = true;
 
         public static final boolean RIGHT_SENSOR_PHASE = true;
         public static final boolean IS_RIGHT_INVERTED = true;
 
+        public static final int JOYSTICK_PORT = 0;
```
Comment (barel13) [inline src/main/java/frc/robot/Ports.java]: You've already defined the port in Controls

File: src/main/java/frc/robot/RobotContainer.java [hunk 21]
```diff
@@ -10,11 +10,14 @@
...
 
 public class RobotContainer {
-  // The robot's subsystems and commands are defined here...
-    public ExampleSubsystem exampleSubsystem = ExampleSubsystem.getInstance();
-    private final XboxController xbox = new XboxController(Ports.Controls.XBOX);
-    private final JoystickButton a = new JoystickButton(xbox, XboxController.Button.kA.value);
+    private GenericHID xbox;
+    //    private final XboxController xbox = new XboxController(Ports.Controls.JOYSTICK);
+   // private final JoystickButton a = new JoystickButton(xbox, XboxController.Button.kA.value);
     private final Climber climber = Climber.getInstance();
```
Comment (barel13) [inline src/main/java/frc/robot/RobotContainer.java]: You don't use this part, delete it

**Review by barel13 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 22]
```diff
@@ -3,79 +3,132 @@
...
          */
-        leftMotor.setSensorPhase(Ports.Climber.LEFT_SENSOR_PHASE);
+        mainMotor.setNeutralMode(NeutralMode.Brake);
 
         /*
-         checking is motor inverted.
+        sets the phase of the sensor.
          */
-        leftMotor.setInverted(Ports.Climber.IS_LEFT_INVERTED);
+        mainMotor.setSensorPhase(Ports.Climber.MAIN_SENSOR_PHASE);
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: I only found out today, but you don't need to set the sensor phase of Falcons ([reference](https://phoenix-documentation.readthedocs.io/en/latest/ch14_MCSensor.html#sensor-phase)).
Also, I recommend using `TalonFXInvertType`  for the `setInverted` as it's more descriptive (I only found out about it today)

**Review by barel13 — APPROVED** [review]

**Comment by adamv17** [comment]
I approve

**Pull request merged**